src/Tasks/ResolveAssemblyReferences/Server/RarController.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.IO.Pipes;
+using System.Security.AccessControl;
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Contract;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Services;
+using Microsoft.VisualStudio.Threading;
+
+using StreamJsonRpc;
+
+#nullable enable
+namespace Microsoft.Build.Tasks.ResolveAssemblyReferences.Server
+{
+    public sealed class RarController
+    {
+        private const int PipeBufferSize = 131072;
+
+        /// <summary>
+        /// Name of <see cref="NamedPipeServerStream"/>
+        /// </summary>
+        private readonly string _pipeName;
+
+        /// <summary>
+        /// Handler for all incoming tasks
+        /// </summary>
+        private readonly IResolveAssemblyReferenceTaskHandler _resolveAssemblyReferenceTaskHandler;
+
+        /// <summary>
+        /// Timeout for incoming connections
+        /// </summary>
+        private readonly TimeSpan Timeout = new TimeSpan(0, 15, 0);
+
+        public RarController(string pipeName, TimeSpan? timeout = null) : this(pipeName, timeout: timeout, resolveAssemblyReferenceTaskHandler: new RarTaskHandler())
+        {
+        }
+
+        internal RarController(string pipeName, IResolveAssemblyReferenceTaskHandler resolveAssemblyReferenceTaskHandler, TimeSpan? timeout = null)
+        {
+            _pipeName = pipeName;
+            _resolveAssemblyReferenceTaskHandler = resolveAssemblyReferenceTaskHandler;
+
+            if (timeout.HasValue)
+            {
+                Timeout = timeout.Value;
+            }
+        }
+
+        public async Task<int> StartAsync(CancellationToken cancellationToken = default)
+        {
+            using ServerMutex mutex = new ServerMutex(_pipeName);
+
+            if (!mutex.IsLocked)
+            {
+                return 1;
+            }
+
+            using CancellationTokenSource cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            CancellationToken token = cancellationTokenSource.Token;
+
+            while (!token.IsCancellationRequested)
```

Comment:
Should this be checking `cancellationTokenSource` instead of `token` to also cover the timeout cancellation?

Reply:
Not sure if I understand this. `token` should be canceled if  `cancellationTokenSource` is canceled from timeout. Correct me if not.

Reply:
I am not very familiar with .NET async so take with a grain of salt. I am looking at this this statement in the `CreateLinkedTokenSource` docs:

_"Creates a CancellationTokenSource that will be in the canceled state when any of the source tokens in the specified array are in the canceled state."_

and

__tokens
CancellationToken[]
An array that contains the cancellation token instances to observe.__

From this it looks as though the cancellation flows from tokens to the token source.

Reply:
That is true, but I think my application is still valid:
**cancellationToken** - normal token
 **cancellationTokenSource** - source of tokens which is canceled if: a) `cancellationToken` is canceled or b) if the timeout expires.
**token** - token from `cancellationTokenSource` which will have property `IsCancellationRequested` set to true if any of the conditions in the `cancellationTokenSource` is fulfilled.

Reply:
Sorry! I totally misread it, please ignore me!